api/nodes/serializers.py

```diff
@@ -485,58 +485,88 @@ def get_absolute_url(self, obj):
+        for inst_id in diff['add']:
+            inst = Institution.load(inst_id)
+            if not inst:
+                raise exceptions.NotFound(detail='Institution with id "{}" was not found'.format(inst_id))
+            insts_to_add.append(inst)
+
+        return insts_to_add, diff['remove'].values()
 
-    def get_related_link(self, obj):
-        return obj.institution_url()
+    def make_instance_obj(self, obj):
+        return {
+            'data': obj.affiliated_institutions,
+            'self': obj
+        }
 
     def update(self, instance, validated_data):
-        node = instance
+        node = instance['self']
         user = self.context['request'].user
 
-        inst = validated_data.get('institution_id', None)
-        if inst:
-            inst = Institution.load(inst)
-            if not inst:
-                raise exceptions.NotFound
-            try:
-                node.add_primary_institution(inst=inst, user=user)
-            except UserNotAffiliatedError:
-                raise exceptions.ValidationError(detail='User not affiliated with institution')
-            node.save()
-            return node
-        node.remove_primary_institution(user)
+        add, remove = self.get_institutions_to_add_remove(
+            institutions=node.affiliated_institutions,
+            new_institutions=validated_data['data']
+        )
+
+        for inst in add:
+            if inst not in user.affiliated_institutions:
+                raise exceptions.PermissionDenied
+
+        for inst in remove:
+            node.remove_affiliated_institution(inst, user)
+        for inst in add:
+            node.add_affiliated_institution(inst, user)
         node.save()
-        return node
 
-    def to_representation(self, obj):
-        data = {}
-        meta = getattr(self, 'Meta', None)
-        type_ = getattr(meta, 'type_', None)
-        assert type_ is not None, 'Must define Meta.type_'
-        relation_id_field = self.fields['id']
-        data['data'] = None
-        if obj.primary_institution:
-            attribute = obj.primary_institution._id
-            relationship = relation_id_field.to_representation(attribute)
-            data['data'] = {'type': type_, 'id': relationship}
-        data['links'] = {key: val for key, val in self.fields.get('links').to_representation(obj).iteritems()}
+        return self.make_instance_obj(node)
 
-        return data
+    def create(self, validated_data):
+        instance = self.context['view'].get_object()
+        user = self.context['request'].user
+        node = instance['self']
+
+        add, remove = self.get_institutions_to_add_remove(
+            institutions=node.affiliated_institutions,
+            new_institutions=validated_data['data']
+        )
+        if not len(add):
+            raise RelationshipPostMakesNoChanges
+
+        for inst in add:
+            if inst not in user.affiliated_institutions:
+                raise exceptions.PermissionDenied
```

Comment:
Please give this an error message indicating the institution for which the user lacks permission.